File: moderate/double_trouble.pl
use strict;
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    chomp($line);
    my ( $o, $r ) = ( ( length $line ) / 2, 1 );
    for ( my $i = 0 ; $i < $o ; $i++ ) {
        if (
            (
                   substr( $line, $i, 1 ) eq "A"
                && substr( $line, $i + $o, 1 ) eq "B"
            )
            || (   substr( $line, $i, 1 ) eq "B"
                && substr( $line, $i + $o, 1 ) eq "A" )
          )
        {
            $r = 0;
            last;
        }
        if (   substr( $line, $i, 1 ) eq "*"
            && substr( $line, $i + $o, 1 ) eq "*" )
        {
            $r <<= 1;
        }
    }
    print "$r\n";
}
close(INFILE);

File: moderate/double_trouble.rb
File.open(ARGV[0]).each_line do |s|
  r = 1
  n = s.length / 2
  (0..n).each do |i|
    if (s[i] == 'A' && s[i + n] == 'B') || (s[i] == 'B' && s[i + n] == 'A')
      r = 0
      break
    end
    r *= 2 if s[i] == '*' && s[i + n] == '*'
  end
  puts r
end
